src/main/java/com/artozersky/HackerNewsAPI/service/impl/NewsPostServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+import com.artozersky.HackerNewsAPI.repository.PostRepository;
+import com.artozersky.HackerNewsAPI.service.NewsPostService;
+import java.util.stream.Collectors;
+
+import jakarta.validation.Valid;
+
+@Service
+public class NewsPostServiceImpl implements NewsPostService {
+    
+    @Autowired
+    private PostRepository postRepository;
+    
+    @Autowired
+    private ModelMapper modelMapper;
+
+    @Autowired
+    private CacheEntityServiceImpl cacheService;
+
+    private Integer limit = 400;
+
+    private final int cacheSize;
+
+    @Autowired
+    public NewsPostServiceImpl(@Value("${cache.size:100}") int cacheSize, @Value("${posts.page.limit:400}") int limit) {
+        this.cacheSize = cacheSize;
+        this.limit = limit;
+        this.cacheService = new CacheEntityServiceImpl(cacheSize);
+    }
+    
+    @Override
+    public  NewsPostsResponseDTOImpl getPostById(Long postId) {
+
+        NewsPostModelImpl cachedValue = cacheService.getPostFromCacheById(postId);
+
+        if (cachedValue != null) {
+            return convertToDTO(cachedValue);
+        }
+
+        NewsPostModelImpl postModel = postRepository.findById(postId)
+            .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+
+        //cacheService.putPostInCache(postModel);
+        //should i update the cache with the top posts???
+
+        return convertToDTO(postModel);
+    }
+
+    @Override
+    public List<NewsPostsResponseDTOImpl> getAllPosts(int limit) {
+
+        if (limit <= 0) {
+            throw new CustomServiceException("Limit must be greater than 0");
+        }
+
+        Pageable pageable = PageRequest.of(0, limit);
+        Page<NewsPostModelImpl> page = postRepository.findAll(pageable);
+
+        return page.getContent().stream()
+                .map(this::convertToDTO)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<NewsPostsResponseDTOImpl> getTopPosts(int limit) {
+
+    List<NewsPostModelImpl> cachedPosts = cacheService.getAllPostsFromCache();//stale will be checked inside
+
+    int postsNeeded = limit - cachedPosts.size();
+
+    System.out.println("cached posts size: " + cachedPosts.size() );
+
+    System.out.println("posts needed: " + postsNeeded );
+
+    List<NewsPostModelImpl> dbPosts = new ArrayList<>();
+
+    dbPosts = fetchPostsFromDbIfNeeded(postsNeeded, cachedPosts);
+
+    System.out.println("fetched from db: " + dbPosts.size() );
+
+    List<NewsPostModelImpl> combinedPosts = new ArrayList<>(cachedPosts);
+    combinedPosts.addAll(dbPosts);
+
+    return combinedPosts.stream()
+        .map(this::convertToDTO)
+        .collect(Collectors.toList());
+}
+
+    
+    @Override
+    public NewsPostsResponseDTOImpl savePost(@Valid NewsPostsCreateDTOImpl postCreateDTO) {
+        
+        NewsPostModelImpl post = modelMapper.map(postCreateDTO, NewsPostModelImpl.class);
+        
+        post.initialize();
```

Comment:
as said in the last review internally and dynamically the post should init its fields like the score calc time elapsed time of creation and so on there are annotation for it that will trigger every time a new object (post model) is created it will init its fields according to some functions. read about it. post.initialize looks weird here.